subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,792 @@
+  // avoid waiting for the next NALU.
+  if (last_nalu) {
+    gboolean au_alignment = rtph266pay->alignment == ALIGNMENT_AU;
+    gboolean marker =
+        GST_BUFFER_FLAG_IS_SET (last_nalu->nalu_buf, GST_BUFFER_FLAG_MARKER);
+    last_nalu->au_end = last_nalu->au_end || au_alignment || marker;
+  }
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_push_pending_data (GstRtpH266Pay * rtph266pay, gboolean eos)
+{
+  GQueue *nalus = &rtph266pay->nalus;
+  GstFlowReturn ret = GST_FLOW_OK;
+  Nalu *nalu;
+
+  // TODO: handle AP
+  // TODO: send XPS if needed
+
+  // TODO: Maybe it's worth to avoid iterating the NALU list twice, maybe not
+  gst_rtp_h266_pay_set_au_boundaries (rtph266pay);
+
+  // Try to push all NALUs
+  while ((nalu = g_queue_pop_head (nalus))) {
+    gboolean is_last = g_queue_get_length (nalus) == 0;
+
+    // Can't push the last NALU without knowing if it's the end of an AU,
+    // because setting the M bit could be necessary. But have to push it if
+    // we're on EOS
+    if (!eos && is_last && !nalu->au_end) {
+      GST_DEBUG_OBJECT (rtph266pay, "Keeping last NALU: %" NALU_PTR_FORMAT,
+          NALU_ARGS (nalu));
+      g_queue_push_head (nalus, nalu);
+      break;
+    }
+
+    if (gst_rtp_h266_pay_has_to_send_xps (rtph266pay, nalu))
+      gst_rtp_h266_pay_queue_xps_cache (rtph266pay);
+
+    GST_DEBUG_OBJECT (rtph266pay, "Pushing NALU: %" NALU_PTR_FORMAT,
+        NALU_ARGS (nalu));
+
+    if (gst_rtp_h266_pay_up_fits_in_mtu (rtph266pay, nalu))
+      ret = gst_rtp_h266_pay_push_up (rtph266pay, nalu);
+    else
+      ret = gst_rtp_h266_pay_push_fu (rtph266pay, nalu);
+
+    nalu_free (nalu);
+    if (ret != GST_FLOW_OK)
+      break;
+  }
+
+  return ret;
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_push_up (GstRtpH266Pay * rtph266pay, const Nalu * nalu)
+{
+  GstRTPBasePayload *rtpbasepay = GST_RTP_BASE_PAYLOAD (rtph266pay);
+  GstRTPBuffer rtp = GST_RTP_BUFFER_INIT;
+  GstBuffer *out_buf;
+
+  // Allocate just the RTP header. We'll add the buffer payload later. This way
+  // we avoid unnecessary copies
+  out_buf = gst_rtp_base_payload_allocate_output_buffer (rtpbasepay, 0, 0, 0);
+  if (!gst_rtp_buffer_map (out_buf, GST_MAP_WRITE, &rtp))
+    goto error;
+
+  // Copy buffer metadata
+  gst_buffer_copy_into (out_buf, nalu->nalu_buf,
+      GST_BUFFER_COPY_FLAGS | GST_BUFFER_COPY_TIMESTAMPS, 0, -1);
+  gst_rtp_buffer_set_marker (&rtp, nalu->au_end);
+
+  // Append the payload to the output buffer
+  out_buf = gst_buffer_append (out_buf, gst_buffer_ref (nalu->hdr_buf));
+  // TODO: Conditionally append DONL
+  out_buf = gst_buffer_append (out_buf, gst_buffer_ref (nalu->rbsp_buf));
+
+  gst_rtp_buffer_unmap (&rtp);
+  GST_DEBUG_OBJECT (rtph266pay, "Pushing UP %" GST_PTR_FORMAT, out_buf);
+  return gst_rtp_base_payload_push (rtpbasepay, out_buf);
+error:
+  gst_buffer_unref (out_buf);
+  return GST_FLOW_ERROR;
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_push_fu (GstRtpH266Pay * rtph266pay, const Nalu * nalu)
```

Comment:
I would use something like "fragment_unit" or similar instead of `fu`. I think it easier to understand what it will do [*](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gst-plugins-good/gst/rtp/gstrtph265pay.c#L1174)

Reply:
The standard refers to it as `FU` and the name is already too long.

Reply:
changed in #42